pkg/hintrunner/zero/hintparser.go

```diff
@@ -122,80 +148,179 @@ func (expression DerefCastExp) Evaluate() (hinter.Reference, error) {
+		// so the expression has to follow the pattern:
+		// reg + off + off + ... + off
+		for _, term := range expression.AddExp {
+			rightExp, err := term.TermExp.Evaluate()
+			if err != nil {
+				return nil, err
+			}
+			rightResult, ok := rightExp.(hinter.Immediate)
+			if !ok {
+				return nil, fmt.Errorf("invalid arithmetic expression")
+			}
+
+			off, ok := utils.Int16FromFelt((*fp.Element)(&rightResult))
+			if !ok {
+				return nil, fmt.Errorf("invalid arithmetic expression")
+			}
+
+			if term.Operator == "-" {
+				off = -off
+			}
+
+			switch cellRef := leftResult.(type) {
+			case hinter.ApCellRef:
+				oldOffset := int16(cellRef)
+				leftResult = hinter.ApCellRef(off + oldOffset)
+				continue
+			case hinter.FpCellRef:
+				oldOffset := int16(cellRef)
+				leftResult = hinter.FpCellRef(off + oldOffset)
+				continue
+			}
+		}
+		return leftResult, nil
+	} else {
+		for _, term := range expression.AddExp {
+			rightExp, err := term.TermExp.Evaluate()
+			if err != nil {
+				return nil, err
+			}
+
+			op, err := parseOperator(term.Operator)
+			if err != nil {
+				return nil, err
+			}
+
+			if leftResult, ok := leftExp.(hinter.ResOperander); ok {
+				if rightResult, ok := rightExp.(hinter.ResOperander); ok {
+					leftExp = hinter.BinaryOp{
+						Operator: op,
+						Lhs:      leftResult,
+						Rhs:      rightResult,
+					}
+					continue
+				}
+			}
+			return nil, fmt.Errorf("invalid arithmetic expression")
+		}
+		return leftExp, nil
+	}
+
+}
+
+func (expression TermExp) Evaluate() (hinter.Reference, error) {
+	switch {
+	case expression.ProdExp != nil:
+		return expression.ProdExp.Evaluate()
+	case expression.Exp != nil:
+		return expression.Exp.Evaluate()
 	default:
 		return nil, fmt.Errorf("unexpected identifier value")
 	}
 }
 
-func (expression Expression) Evaluate() (any, error) {
+func (expression ProdExp) Evaluate() (hinter.Reference, error) {
+	leftExp, err := expression.LeftExp.Evaluate()
+	if err != nil {
+		return nil, err
+	}
+	rightExp, err := expression.RightExp.Evaluate()
+	if err != nil {
+		return nil, err
+	}
+
+	if leftOp, ok := leftExp.(hinter.ResOperander); ok {
+		if rightOp, ok := rightExp.(hinter.ResOperander); ok {
+			return hinter.BinaryOp{
+				Operator: hinter.Mul,
+				Lhs:      leftOp,
+				Rhs:      rightOp,
+			}, nil
+		}
+	}
+	return nil, fmt.Errorf("unexpected product expression")
+}
+
+func (expression Expression) Evaluate() (hinter.Reference, error) {
 	switch {
+	case expression.IntExp != nil:
+		intExp, err := expression.IntExp.Evaluate()
+		if err != nil {
+			return nil, err
+		}
+		return hinter.Immediate(*new(fp.Element).SetBigInt(intExp)), nil
 	case expression.CellRefExp != nil:
 		return expression.CellRefExp.Evaluate()
 	case expression.DerefExp != nil:
 		return expression.DerefExp.Evaluate()
-	case expression.BinOpExp != nil:
-		return expression.BinOpExp.Evaluate()
+	case expression.DDerefExp != nil:
```

Comment:
I don't quite understand, why could we replace parsing Binary operator with double deref here?

Reply:
So it's not exactly that one is being replaced by the other, the entire grammar changed, so now the Expression type takes the "atomic" expressions of operanders: Immediate, CellRef, Deref and DoubleDeref. For BinOp, notice that with this grammar we can now have several terms in a sumation instead of only 2, so we introduced the type ArithmeticExpression in the grammar, which can be evaluated into Binary Operations. Also, for allowing multiple terms, now the left and right hand side of the binary operation are resOperanders, which allows the structure to be recursive.
For example, the expression: `"cast([ap + 3] + [ap + (-4)] * 18 + 15, felt)"` would be evaluated as:
```
hinter.BinaryOp{
    Operator: hinter.Add,
    Lhs: hinter.BinaryOp{
        Operator: hinter.Add,
        Lhs:      hinter.Deref{
             Deref: hinter.ApCellRef(3),
        },
        Rhs: hinter.BinaryOp{
             Operator: hinter.Mul,
             Lhs: hinter.Deref{
                  Deref: hinter.ApCellRef(-4),
             },
            Rhs: hinter.Immediate(18),
        },
    },
    Rhs: hinter.Immediate(15),
}
```